### DISCUSSION TITLE
Injecting gradient directly into the training extension

### DISCUSSION THREAD

**Discussion by berndporr** [pytorch/executorch#17297] [Q&A]
I do realtime learning with libtorch where the gradient is calculated elsewhere and would like to translate that to executorch: https://github.com/berndporr/dnf_torch this is realtime noise cancellations but I have also other applications where the gradient comes further afield so really not associated with a simple subtraction operation as I also do RL. In libtorch I can directly inject the gradient and then do the backwards pass: https://github.com/berndporr/dnf_torch/blob/main/dnf_torch.cpp#L99 
```
torch::Tensor gradient = torch::tensor({-f_nn}).to(device);
output.backward(gradient);
optimizer.step();
```
However, in executorch forward/backward is in one go. How do I go about this?
Looking at the XOR example I could inject the gradient as an additional input? Roughly:
```
class TrainingNet(nn.Module):
    def __init__(self, net):
        super().__init__()
        self.net = net

    def forward(self, gradient, input):
        pred = self.net(input)
        return gradient, pred.detach().argmax(dim=1)
```
and then lower that and inject the gradient as the first input. Would that be the way to go?

**Comment by cccclai** [comment]
@JacobSzwejbka

**Comment by berndporr** [comment]
Stupid me. loss != gradient. So we need to square the gradient and then that's un-squared during exporting to calc the grad again:
```
class GradientLoss(nn.Module):
    def forward(self, grad):
        loss = torch.square(grad);
        return loss

class TrainingNet(nn.Module):
    def __init__(self, net):
        super().__init__()
        self.net = net
        self.criterion = GradientLoss()

    def forward(self, noiseRef, noisySignal):
        remover = self.net(noiseRef)
        grad = noisySignal - remover
        loss = self.criterion(grad)
        return loss, remover.detach()
```
I'm not sure why this works as it should need the output of the network in the derivative:
```
class GradientLoss(nn.Module):
    def forward(self, grad, output):
        loss = torch.square(output * grad);
        return loss

class TrainingNet(nn.Module):
    def __init__(self, net):
        super().__init__()
        self.net = net
        self.criterion = GradientLoss()

    def forward(self, noiseRef, noisySignal):
        remover = self.net(noiseRef)
        grad = noisySignal - remover
        loss = self.criterion(grad,remover)
        return loss, remover.detach()
```
The eagle eyed person of course spots that for this realtime noise cancellation paradigm I can just actually create a bog standard loss and that works as well:
```
class TrainingNet(nn.Module):
    def __init__(self, net):
        super().__init__()
        self.net = net
        self.criterion = nn.MSELoss()

    def forward(self, noiseRef, noisySignal):
        remover = self.net(noiseRef)
        loss = self.criterion(remover,noisySignal)
        return loss, remover.detach()
```
as done here: https://github.com/berndporr/dnf_executorch/ (work in progress).
But for other RL stuff I do I "only" have the gradient but I guess that's generally the way to do it?

**Comment by berndporr** [comment]
To answer my own question a bit further. The "Gradient Loss" works because it's dependent on the "remover" here. I'll stick to this which then allows to use the "gradient" or "error" in LMS filter terminology also to be returned to C++ as it's the filter's output:
```
# DNF encoder
class Net(nn.Module):
    def __init__(self,nTaps,nLayers,nonlin):
        super().__init__()

        torch.manual_seed(42)
        self.seq = nn.Sequential()
        b = np.exp(np.log(nTaps)/nLayers);
        nInput = nTaps
        for i in range(nLayers):
            if (i == (nLayers-1)):
                nOutput = 1
            else:
                nOutput = int(np.ceil(nTaps / np.pow(b,i+1)));
            print("Created layer",i,"with",nInput,"->",nOutput)
            l = nn.Linear(nInput, nOutput)
            nn.init.xavier_uniform(l.weight)
            nn.init.zeros_(l.bias)
            self.seq.add_module("Layer"+str(i),l)
            self.seq.add_module("Nonlin"+str(i),nonlin)
            nInput = nOutput

    def forward(self, x):
        return self.seq(x)

class GradientLoss(nn.Module):
    def forward(self, grad):
        loss = torch.square(grad);
        return loss

class TrainingNet(nn.Module):
    def __init__(self, net):
        super().__init__()
        self.net = net
        self.criterion = GradientLoss()

    def forward(self, noiseRef, noisySignal):
        remover = self.net(noiseRef)
        error = noisySignal - remover
        loss = self.criterion(error)
        return loss, remover.detach(), error.detach()
```
Full implementation here: https://github.com/berndporr/dnf_executorch
